### PR TITLE
core/chains/terra/monitor: add Terra BalanceMonitor

### PR DESCRIPTION
**Pull request by jmank88**
https://app.shortcut.com/chainlinklabs/story/26747

### CODE REVIEW

File: core/chains/terra/monitor/balance.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	}
+}
+
+func (b *balanceMonitor) updateBalances() {
+	keys, err := b.ks.GetAll()
+	if err != nil {
+		b.lggr.Errorw("Failed to get keys", "err", err)
+		return
+	}
+	var rw client.ReaderWriter
```
Comment (pinebit) [inline core/chains/terra/monitor/balance.go:90]: How is this initialized here?

File: core/chains/terra/monitor/balance.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	for _, k := range keys {
+		select {
+		case <-b.stop:
+			return
+		default:
+		}
+		acc := sdk.AccAddress(k.PublicKey().Address())
+		bal, err := rw.Balance(acc, "uluna") //TODO async to prevent blocking shutdown?
```
Comment (pinebit) [inline core/chains/terra/monitor/balance.go:98]: This should be receiving a context, internally it is using `context.Background()` 😓

File: core/chains/terra/monitor/balance.go (hunk 1 above)
Comment (jmank88) [reply core/chains/terra/monitor/balance.go:90]: Still incomplete :factory_worker:

File: core/chains/terra/monitor/balance.go (hunk 2 above)
Comment (jmank88) [reply core/chains/terra/monitor/balance.go:98]: IIRC somewhere even further down doesn't respect the context, and we've had to work around it already. I'll revisit this though to be sure.

File: core/chains/terra/chain.go [hunk 3]
```diff
@@ -67,7 +70,8 @@ func NewChain(db *sqlx.DB, ks keystore.Terra, logCfg pg.LogConfig, eb pg.EventBr
 			}, nil
 		}),
 	}, lggr)
-	ch.txm = terratxm.NewTxm(db, tc, *gpe, dbchain.ID, cfg, ks, lggr, logCfg, eb)
+	ch.txm = terratxm.NewTxm(db, tc, *gpe, ch.id, cfg, ks, lggr, logCfg, eb)
+	ch.balanceMonitor = monitor.NewBalanceMonitor(ch.id, cfg, lggr, ks, ch.Reader)
```
Comment (jmank88) [inline core/chains/terra/chain.go:74]: Should this be optional, like EVM? e.g. `cfg.BalanceMonitorEnabled()` - if so, default on or off?

File: core/chains/terra/monitor/balance.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	}
+	var gotBals int
+	for _, k := range keys {
+		select {
+		case <-b.stop:
+			return
+		default:
+		}
+		acc := sdk.AccAddress(k.PublicKey().Address())
+		bal, err := r.Balance(acc, "uluna")
```
Comment (jmank88) [inline core/chains/terra/monitor/balance.go:121]: Should we monitor any other denominations, or just (u)luna?

File: core/chains/terra/monitor/balance.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	if err != nil {
+		b.lggr.Errorw("Failed to convert balance to float", "err", err)
+	}
+	promTerraBalance.WithLabelValues(acc.String(), b.chainID, bal.GetDenom()).Set(balF)
+}
+
+func (b *balanceMonitor) monitor() {
+	defer close(b.done)
+
+	tick := time.After(utils.WithJitter(b.cfg.BlockRate()))
```
Comment (jmank88) [inline core/chains/terra/monitor/balance.go:75]: Too fast?

Comment (connorwstein) [reply core/chains/terra/monitor/balance.go:75]: I think its ok since this is node wide

File: core/chains/terra/monitor/balance.go (hunk 4 above)
Comment (connorwstein) [reply core/chains/terra/monitor/balance.go:121]: just uluna to start seems fine. I think if/when we support multiple denominations we can make that change globally

File: core/chains/terra/monitor/balance.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+		return
+	}
+	r, err := b.getReader()
+	if err != nil {
+		b.lggr.Errorw("Failed to get client", "err", err)
+		return
+	}
+	var gotBals int
+	for _, k := range keys {
+		select {
```
Comment (connorwstein) [inline core/chains/terra/monitor/balance.go:115]: so with this the worst case shutdown time is the length of a r.Balance call right? worth a comment

File: core/chains/terra/monitor/balance.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+		}
+		gotBals++
+		balLuna, err := sdk.ConvertDecCoin(sdk.NewDecCoinFromCoin(*bal), "luna")
+		if err != nil {
+			b.lggr.Errorw("Failed to convert uluna to luna", "account", acc, "err", err)
+			continue
+		}
+		b.updateProm(acc, &balLuna)
+	}
+	if gotBals == 0 {
```
Comment (connorwstein) [inline core/chains/terra/monitor/balance.go:134]: could just use a bool?

File: core/chains/terra/monitor/balance.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
+package monitor
```
Comment (connorwstein) [inline core/chains/terra/monitor/balance.go:1]: could rename the evm one too for consistency

File: core/chains/terra/chain.go (hunk 3 above)
Comment (connorwstein) [reply core/chains/terra/chain.go:74]: seems fairly lightweight and beneficial I'm not sure why you'd want to disable it. We don't let you disable any other prom metrics afaik, I'd argue we should just remove the eth flag

**Review by connorwstein — COMMENTED** [review]
Any easy way to add some test coverage?

File: core/chains/terra/monitor/prom.go [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var promTerraBalance = promauto.NewGaugeVec(
+	prometheus.GaugeOpts{Name: "terra_balance", Help: "Terra account balances"},
+	[]string{"account", "terraChainID", "denomination"},
+)
+
+func (b *balanceMonitor) updateProm(acc sdk.AccAddress, bal *sdk.DecCoin) {
```
Comment (jmank88) [inline core/chains/terra/monitor/prom.go:14]: This is not covered by the test. How do we normally test prometheus?

Comment (connorwstein) [reply core/chains/terra/monitor/prom.go:14]: normally just in QA / locally running a node if possible. I think its ok to leave this to QA

**Review by connorwstein — APPROVED** [review]

**Pull request merged**